From f06753afe2013023c71dd8403f837553e6cefbd0 Mon Sep 17 00:00:00 2001
From: Zhiqiang Ni <nizhiqiang1@huawei.com>
Date: Thu, 19 Sep 2024 16:29:19 +0800
Subject: [PATCH 464/960] downstream: KVM: arm64: Add a kvm parameter to
 control guest wfi trapping

Currently, on GICv4+ system if a vCPU is only runnable process
on a pCPU and we are using direct injection of interrupts, the
guest WFI trapping will be turn off. So in host view, the vcpu
process's CPU util is almost 100%, which may confuse someone.
Let's add a kvm parameter to control guest wfi trapping.

Signed-off-by: Zhiqiang Ni <nizhiqiang1@huawei.com>
Signed-off-by: chenyi <chenyi211@h-partners.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 arch/arm64/include/asm/kvm_emulate.h |  3 ++
 arch/arm64/include/asm/kvm_host.h    |  3 ++
 arch/arm64/kvm/arm.c                 | 49 ++++++++++++++++++++++++++++
 3 files changed, 55 insertions(+)

diff --git a/arch/arm64/include/asm/kvm_emulate.h b/arch/arm64/include/asm/kvm_emulate.h
index 2798f4f1f4c8..9579ab0646a9 100644
--- a/arch/arm64/include/asm/kvm_emulate.h
+++ b/arch/arm64/include/asm/kvm_emulate.h
@@ -116,6 +116,9 @@ static inline void vcpu_clear_wfx_traps(struct kvm_vcpu *vcpu)
 		vcpu->arch.hcr_el2 &= ~HCR_TWI;
 	else
 		vcpu->arch.hcr_el2 |= HCR_TWI;
+
+	if (force_wfi_trap)
+		vcpu->arch.hcr_el2 |= HCR_TWI;
 }
 
 static inline void vcpu_set_wfx_traps(struct kvm_vcpu *vcpu)
diff --git a/arch/arm64/include/asm/kvm_host.h b/arch/arm64/include/asm/kvm_host.h
index 8a0cebf24d31..e267c6e39b4f 100644
--- a/arch/arm64/include/asm/kvm_host.h
+++ b/arch/arm64/include/asm/kvm_host.h
@@ -50,6 +50,7 @@
 #define KVM_REQ_RELOAD_PMU	KVM_ARCH_REQ(5)
 #define KVM_REQ_SUSPEND		KVM_ARCH_REQ(6)
 #define KVM_REQ_RESYNC_PMU_EL0	KVM_ARCH_REQ(7)
+#define KVM_REQ_RELOAD_WFI_TRAPS	KVM_ARCH_REQ(9)
 
 #define KVM_DIRTY_LOG_MANUAL_CAPS   (KVM_DIRTY_LOG_MANUAL_PROTECT_ENABLE | \
 				     KVM_DIRTY_LOG_INITIALLY_SET)
@@ -1190,4 +1191,6 @@ static inline void kvm_hyp_reserve(void) { }
 void kvm_arm_vcpu_power_off(struct kvm_vcpu *vcpu);
 bool kvm_arm_vcpu_stopped(struct kvm_vcpu *vcpu);
 
+extern bool force_wfi_trap;
+
 #endif /* __ARM64_KVM_HOST_H__ */
diff --git a/arch/arm64/kvm/arm.c b/arch/arm64/kvm/arm.c
index d88547e00bd5..b35c69330c98 100644
--- a/arch/arm64/kvm/arm.c
+++ b/arch/arm64/kvm/arm.c
@@ -63,6 +63,48 @@ bool is_kvm_arm_initialised(void)
 	return kvm_arm_initialised;
 }
 
+static int vcpu_req_reload_wfi_traps(const char *val, const struct kernel_param *kp);
+
+static const struct kernel_param_ops force_wfi_trap_ops = {
+	.set = vcpu_req_reload_wfi_traps,
+	.get = param_get_bool,
+};
+
+bool force_wfi_trap;
+module_param_cb(force_wfi_trap, &force_wfi_trap_ops, &force_wfi_trap, 0644);
+
+static int vcpu_req_reload_wfi_traps(const char *val, const struct kernel_param *kp)
+{
+	struct kvm *kvm;
+	bool oldvalue;
+	int err;
+
+	oldvalue = force_wfi_trap;
+	err = param_set_bool(val, kp);
+	if (err)
+		return err;
+
+	if (oldvalue == force_wfi_trap)
+		return err;
+
+	/*
+	 * If set the force_wfi_trap from 1 to 0, no need to kick vcpus here.
+	 * The HCR_TWI flag will be cleared in kvm_arch_vcpu_load().
+	 */
+	if (force_wfi_trap == 0)
+		return 0;
+
+	/*
+	 * We need to kick vcpus out of guest mode here to reload
+	 * wfx trapping config when re-enter guest mode.
+	 */
+	mutex_lock(&kvm_lock);
+	list_for_each_entry(kvm, &vm_list, vm_list)
+		kvm_make_all_cpus_request(kvm, KVM_REQ_RELOAD_WFI_TRAPS);
+	mutex_unlock(&kvm_lock);
+	return err;
+}
+
 int kvm_arch_vcpu_should_kick(struct kvm_vcpu *vcpu)
 {
 	return kvm_vcpu_exiting_guest_mode(vcpu) == IN_GUEST_MODE;
@@ -800,6 +842,13 @@ static int check_vcpu_requests(struct kvm_vcpu *vcpu)
 
 		if (kvm_dirty_ring_check_request(vcpu))
 			return 0;
+
+		if (kvm_check_request(KVM_REQ_RELOAD_WFI_TRAPS, vcpu)) {
+			if (single_task_running())
+				vcpu_clear_wfx_traps(vcpu);
+			else
+				vcpu_set_wfx_traps(vcpu);
+		}
 	}
 
 	return 1;
-- 
2.43.0

